feat(SPSTRAT-465): add task for marketplacesvm

JAVGan opened the pull request:
This commit introduces a new task named `marketplaces-push-disk-images` which will be used to deliver disk images to various cloud marketplaces using the `marketplacesvm_push_wrapper`.

Refers to [SPSTRAT-465](https://issues.redhat.com//browse/SPSTRAT-465)

JAVGan committed:
feat(SPSTRAT-465): add task for marketplacesvm

This commit introduces a new task named
`marketplacesvm-push-disk-images` which will be used to
 deliver disk images to various cloud marketplaces
using the `marketplacesvm_push_wrapper`.

Signed-off-by: Jonathan Gangi <jgangi@redhat.com>

JAVGan commented:
@lslebodn FYI moved here.

@johnbieren is there a way to test that?

lslebodn commented on internal/tasks/marketplaces-push-disk-images-task.yaml:
```diff
@@ -0,0 +1,149 @@
+                if [ "${FILENAME##*\.}" = "vhd" ];
+                    image_type="VHD"
+                else
+                    image_type="AMI"
```
I would check the `raw` extension here for AWS/AMI
neither qcow2 nor iso are supported. (In another words; we should not automatically fallback to AMI

JAVGan replied:
Good catch!

I've fixed that as well as added a directory for the task with a `README.md` file so it is not compliant with the other tasks.

PTAL again and let me know if you find something else

lslebodn reviewed (commented):
Can we somewhere in the code validate staging structure with pushsource-ls
as we discussed in https://github.com/konflux-ci/release-service-utils/pull/326#discussion_r1860795719?

mmalina commented on internal/tasks/marketplacesvm-push-disk-images-task/marketplaces-push-disk-images-task.yaml:
We'll need some tests for this task. Also, you will need to add a symlink to this file to the internal/resources directory if you want to get this deployed to the app sre cluster (for internal-request usage).

JAVGan replied:
Question: Do we need to have it internal?

I've made like this because I'm uncertain about `oras pull` and followed the example from `pulp` as well. But we don't need to call `Exodus` nor any other service..

JAVGan replied:
Adding @johnbieren here as well so he may have an answer for this question

mmalina replied:
Ha, good point. I saw this being added in internal and I though: Right, this will be done as InternalRequest.

But now I recall our discussion in the call we had a few weeks ago and I think you said that everything you access is in the public internet, right? So you don't need access to anything in Red Hat private network.

In that case yes, this should be a normal task in the `tasks/` directory.

johnbieren replied:
If it doesn't need to access things inside the firewall, we'd prefer it not being an internal task. +1 to Martin's reply

mmalina commented on internal/tasks/marketplacesvm-push-disk-images-task/marketplaces-push-disk-images-task.yaml:
```diff
@@ -0,0 +1,151 @@
+        RUNNING_JOBS="\j" # Bash parameter for number of jobs currently running
+        NUM_COMPONENTS=$(jq '.components | length' <<< "$SNAPSHOT_JSON")
+
+        process_component() { # Expected argument is [component json]
```
So this will pull the image and prepare the data for the push, right? And then the push will always happen all in one go? So there is no way to make that parallel?

JAVGan replied:
Yes, the `process_component` will be called in parallel to prepare the data and then the push will be executed in one go through the `marketplacesvm_push_wrapper`. However, the tooling is multi-thread so in the end the push execution will be done in parallel

mmalina replied:
OK, thanks. Would it be more appropriate to call the function `pull_component` perhaps? Or `prepare_component`? I think in the other task where this was taken from, the `process_component` function does everything there is to be done - there is no pushing afterwards (but I haven't checked now).

mmalina replied:
Any comments on this?

JAVGan replied:
Oh! I didn't notice this, it got lost in many threads.

@johnbieren WDYT? Should I update?

johnbieren replied:
I don't have an opinion on it. If Martin thinks a different name is better, it probably doesn't hurt

JAVGan replied:
Done!

I've renamed it to `prepare_component`.

I'll mark this as resolved then.

mmalina replied:
ok, thanks!

JAVGan commented:
@mmalina @johnbieren I've added a single test to ensure the task works, but honestly since I'm very new to testing stuff on Konflux I'm really not sure whether I did it correctly or not... What I did was based in the other tasks with some adaptations for our use cases, please let me know what I need to fix/change and I'll do it.

However, by implementing such tests I ended up with the following questions:

1. As asked earlier do we need to have our task as `internal`? We're just using `oras pull` and `select-oci-auth` to deal with `quay.io` and for the rest we don't need anything from within the firewall... If you agree the task can be external I'll move it out from the `internal` directory.
2. By looking at the tests from [push-disk-images](https://github.com/konflux-ci/release-service-catalog/blob/development/tasks/push-disk-images/tests/test-push-disk-images-production.yaml#L21-L58) I noticed you have 2 different JSON files: `snapshot_spec.json` and `data.json` which are [combined here](https://github.com/konflux-ci/release-service-catalog/blob/development/tasks/push-disk-images/push-disk-images.yaml#L39). I assumed the `snapshot_json` was incoming as a whole, like what we discussed before on slack channel and my task assumes it's a single JSON as well... Should I update it to receive two JSONs aside from our `starmap mappings`? It would end up with 3 JSONs for the parameters if yes...
3. Is there a way I can "validate" or "run the tests" in some place before asking you guys to run it on CI? I believe it would be faster to catch errors and fix stuff.

mmalina commented on internal/tasks/marketplacesvm-push-disk-images-task/README.md:
Let's have the discussion here - that way it's a proper thread which is easier for replies.

> 1. As asked earlier do we need to have our task as `internal`? We're just using `oras pull` and `select-oci-auth` to deal with `quay.io` and for the rest we don't need anything from within the firewall... If you agree the task can be external I'll move it out from the `internal` directory.

Yes, as I said in the other thread, it turns out there is no need for InternalRequests here, so please move this to the main `tasks/` directory. That also means that you can access the shared workspace, so you shouldn't pass snapshot as a string, but rather the path to it in the workspace - you can find multiple examples in other tasks.

> 2. By looking at the tests from [push-disk-images](https://github.com/konflux-ci/release-service-catalog/blob/development/tasks/push-disk-images/tests/test-push-disk-images-production.yaml#L21-L58) I noticed you have 2 different JSON files: `snapshot_spec.json` and `data.json` which are [combined here](https://github.com/konflux-ci/release-service-catalog/blob/development/tasks/push-disk-images/push-disk-images.yaml#L39). I assumed the `snapshot_json` was incoming as a whole, like what we discussed before on slack channel and my task assumes it's a single JSON as well... Should I update it to receive two JSONs aside from our `starmap mappings`? It would end up with 3 JSONs for the parameters if yes...

The snapshot is actually a so called mapped snapshot spec - it's the result of apply mapping - it will contain things from the original snapshot merged with the mapping in the data file (what is in the ReleasePlanAdmission) - basically with the target repository added, and also tags.

So it depends on what you need. But now that you have access to the workspace and the files in it, you can read everything from there. But you will still need a separate task like `collect-marketplacesvm-params` that will collect the secret name from the data file, unless it's always the same and we hardcode it. The reason is that you need to mount the secret in the main task so you need to know what it is when the task starts. Again, you can see many examples in the repo - all the collect-* tasks are examples of that.

And since you will already have a collect task, you may want to load everything in it and pass as parameters for the main task - but that really depends on what looks better - if it's too many params, then it's probably still better to load stuff in the main task directly.

> 3. Is there a way I can "validate" or "run the tests" in some place before asking you guys to run it on CI? I believe it would be faster to catch errors and fix stuff.

You can run the tests in a kind cluster locally, it's pretty easy, see here: https://github.com/konflux-ci/release-service-catalog/blob/development/CONTRIBUTING.md#running-tekton-task-tests-manually

JAVGan replied:
Thank you a lot for the detailed explanation! I'll adjust my PR accordingly.

johnbieren replied:
The only secret is the cloud marketplaces one, and that isn't even in the collect task right now, so I commented elsewhere to remove the collect task. But, if you do need the secret and it is defined in the data file as this thread suggests, yes the collect task is the proper way to do it (but I think it is simpler to just output the secret as the result. The other things currently in the collect- task should just be in the main task and not be passed around as results IMO)

johnbieren commented on tasks/collect-marketplacesvm-params/README.md:
```diff
@@ -0,0 +1,11 @@
+## Parameters
+
+| Name     | Description                                                                                                            | Optional | Default value |
+|----------|------------------------------------------------------------------------------------------------------------------------|----------|---------------|
```
Spacing doesn't align with line 11

johnbieren commented on tasks/collect-marketplacesvm-params/README.md:
I am confused as to why we need this as a separate task. It looks to me like you are just extracting values from the data file to use in `marketplacesvm-push-disk-images` task. Typically, we only create `collect-*` tasks if we need to use secrets, as those have to be mounted in the pod at initialization so have to be discovered ahead of time. But it looks like you are just doing this 
```
        PRODUCT_INFO=$(jq -c '.' "$(workspaces.data.path)/$(params.productInfoPath)")
        STARMAP_MAPPING_FILE="$(workspaces.data.path)/$(params.starmapMappingPath)"
```
so couldn't the logic from this task just go in the other task and only have one new task instead of two?

johnbieren replied:
Wrote this before this https://github.com/konflux-ci/release-service-catalog/pull/719#discussion_r1873651052

JAVGan commented:
@mmalina @johnbieren can you please review again and let me know if the tests passes?

I've rewritten it to collect everything from the task itself. The StArMap mappings and ProductInfo will be given in the RPA so I've updated it to collect from the `dataPath`.

Please let me know if my assumptions were correct as well.

For the `snapsjot_json` I've considered it the merged content after the `apply-mappings`, so the RPA may have the `mappings` property as well to create the staging structure.

johnbieren commented on tasks/marketplacesvm-push-disk-images-task/marketplaces-push-disk-images-task.yaml:
```diff
@@ -0,0 +1,168 @@
+apiVersion: tekton.dev/v1
+kind: Task
+metadata:
+  name: marketplaces-push-disk-images
```
Can you drop the `-task` from the dir name and file name so it matches this? I think you got that convention from our internal dir, but the stuff in `tasks` doesn't need it

JAVGan replied:
Renamed

johnbieren commented on tasks/marketplacesvm-push-disk-images-task/README.md:
```diff
@@ -0,0 +1,14 @@
+| ----------------------- | -------------------------------------------------------------------------------------- | -------- | --------------- |
+| snapshotPath            | Path to the JSON string of the mapped snapshot spec in the data workspace.             | No       | -               |
+| dataPath                | Path to the data JSON in the workspace containing the marketplacesvm options to use.   | No       | -               |
+| cloudMarketplacesSecret | Env specific secret containing the marketplaces credentials.                           | No       | -               |
```
I think my comments were confusing. You are going to need that collect task, but just for this, nothing else

JAVGan replied:
For the secret in this case? Is it just that, right?

johnbieren replied:
Correct because it has to be mounted with the pod so it has to be determined before task start

JAVGan replied:
Thanks, I'll add it, just a min

JAVGan replied:
Added

johnbieren commented on tasks/marketplacesvm-push-disk-images-task/tests/pre-apply-task-hook.sh:
```diff
@@ -0,0 +1,11 @@
+yq -i '.spec.steps[0].script = load_str("'$SCRIPT_DIR'/mocks.sh") + .spec.steps[0].script' "$TASK_PATH"
+
+# Create a dummy secret (and delete it first if it exists)
+kubectl delete secret test-secret --ignore-not-found
```
Can you call this like `marketplace-test-secret` or something? We tend to do that to make it clearer

JAVGan replied:
Renamed

johnbieren commented on tasks/marketplacesvm-push-disk-images-task/marketplaces-push-disk-images-task.yaml:
```diff
@@ -0,0 +1,168 @@
+  annotations:
+    tekton.dev/pipelines.minVersion: "0.12.1"
+    tekton.dev/tags: release
+spec:
```
I think you missed defining the data workspace

JAVGan replied:
Thanks! I've added it

johnbieren commented on tasks/marketplacesvm-push-disk-images-task/marketplaces-push-disk-images-task.yaml:
```diff
@@ -0,0 +1,168 @@
+      type: string
+      description: The maximum number of images to be pulled at once.
+      default: 3
+  results:
```
Since this isn't an internal task, we don't really need this. We do this as a trick in our internal task to expose the results, as they run on a separate cluster. But this we will be able to just see the logs

JAVGan replied:
Removed

johnbieren commented on tasks/marketplacesvm-push-disk-images-task/marketplaces-push-disk-images-task.yaml:
```diff
@@ -0,0 +1,168 @@
+
+        STDERR_FILE=/tmp/stderr.txt
+
+        exitfunc() {
```
This entire exitfunc can be removed for same reason as line 29 comment

JAVGan replied:
Removed

johnbieren commented on tasks/marketplacesvm-push-disk-images-task/tests/mocks.sh:
```diff
@@ -0,0 +1,31 @@
+  echo Mock imarketplacesvm_push_wrapper called with: $*
+  echo $* > "$(workspaces.data.path)/mock_wrapper.txt"
+
+  /home/pubtools-marketplacesvm-wrapper/marketplacesvm_push_wrapper "$@" --dry-run
```
this is nice

johnbieren commented on tasks/marketplacesvm-push-disk-images-task/tests/test-marketplacesvm-push-disk-images.yaml:
```diff
@@ -0,0 +1,175 @@
+              #!/usr/bin/env bash
+              set -eux
+
+              mkdir "$(workspaces.data.path)/results"
```
don't believe this is needed or used

JAVGan replied:
Removed

JAVGan commented:
I've pushed again to fix a syntax error caught by Checkton, hope it pass now

johnbieren commented on tasks/marketplacesvm-push-disk-images-task/tests/test-marketplacesvm-push-disk-images.yaml:
```diff
@@ -0,0 +1,175 @@
+                exit 1
+              fi
+
+              echo "Checks passed"
```
not needed, we will see it is green and not red

JAVGan replied:
Removed

johnbieren commented on tasks/marketplacesvm-push-disk-images-task/tests/test-marketplacesvm-push-disk-images.yaml:
```diff
@@ -0,0 +1,175 @@
+                  exit 1
+              fi
+
+              RESULT_FILE="$(params.result)"
```
I think this is also going to end up removed

JAVGan replied:
Removed

johnbieren reviewed (commented):
Your tests are pretty thorough, thanks for taking the time on that. The last comment from me is can you add the new keys to the schema here https://github.com/konflux-ci/release-service-catalog/blob/development/schema/dataKeys.json?
Once that is added and Martin approves, we can merge (given the linting things are fixed too). We do have e2e tests but this being a new task that isn't part of any pipeline yet, they are irrelevant here

JAVGan commented:
> Your tests are pretty thorough, thanks for taking the time on that. The last comment from me is can you add the new keys to the schema here https://github.com/konflux-ci/release-service-catalog/blob/development/schema/dataKeys.json? Once that is added and Martin approves, we can merge (given the linting things are fixed too). We do have e2e tests but this being a new task that isn't part of any pipeline yet, they are irrelevant here

I've added it.

@mmalina PTAL

mmalina commented on schema/dataKeys.json:
```diff
@@ -437,6 +437,50 @@
+                    "type": "string",
+                    "description": "productVersionName in content gateway"
+                  }
+              },
```
The json is malformed. It seems the closing `}` is missing for `"properties"` here. Not sure if anything else.

JAVGan replied:
I'm going to fix it, just a sec

JAVGan replied:
Fixed, please try running again the CI

mmalina replied:
Looks good now.

mmalina commented on schema/dataKeys.json:
The `cloudMarketplacesSecret` field is missing in the schema file. Also, I was wondering if there is a better place to put it than .mapping.cloudMarketplacesSecret. But I guess it's probably good enough - unless it makes sense to add any to any other existing top level objects like cdn.

JAVGan replied:
Added it

mmalina commented on schema/dataKeys.json:
```diff
@@ -437,6 +437,54 @@
+              },
+              "cloudMarketplacesSecret": {
+                "type": "string",
+                "name": "Secret for cloud marketplaces"
```
This should be description, not name, right?

JAVGan replied:
oops

JAVGan replied:
good catch! Changed that

JAVGan commented:
@mmalina please run the CI again, there was a typo in one of the tests which caused the failure. It should pass now

johnbieren commented on schema/dataKeys.json:
```diff
@@ -437,6 +437,55 @@
+              },
+              "cloudMarketplacesSecret": {
+                "type": "string",
+                "description": "Secret for cloud marketplaces"
```
You index this as mapping.cloudMarketplacesSecret in the task (not sure why it is in mapping tbh) but here you have it as mapping.components[0].cloudMarketplacesSecret

mmalina replied:
Good catch about it being inside components. Regarding it being in mapping, I asked the same thing here: https://github.com/konflux-ci/release-service-catalog/pull/719#discussion_r1875903632

But maybe it's ok - we do have registrySecret there as well.

JAVGan replied:
Oh, let me fix that

JAVGan replied:
Fixed, hope it works now

johnbieren reviewed (commented):
You have a typo in your directory name. 
`tasks/marketplacesvm-push-disk-images/marketplaces-push-disk-images.yaml`
You should have `vm` in both or neither

johnbieren commented on tasks/marketplacesvm-push-disk-images/marketplacesvm-push-disk-images.yaml:
```diff
@@ -0,0 +1,147 @@
+apiVersion: tekton.dev/v1
+kind: Task
+metadata:
+  name: marketplaces-push-disk-images
```
This should align as well so you went with keeping the vm so this should be `marketplacesvm-push-disk-images`

JAVGan replied:
Changed

johnbieren commented on tasks/marketplacesvm-push-disk-images/tests/test-marketplacesvm-push-disk-images.yaml:
```diff
@@ -0,0 +1,172 @@
+          workspace: tests-workspace
+
+    - name: check-result
+      params:
```
no need for param anymore (and it doesn't exist)

JAVGan replied:
Oh, that's right! I got rid of it, thanks!

johnbieren commented on tasks/marketplacesvm-push-disk-images/tests/test-marketplacesvm-push-disk-images.yaml:
```diff
@@ -0,0 +1,172 @@
+
+    - name: run-task
+      taskRef:
+        name: marketplaces-push-disk-images
```
will end up needing the `vm` added

JAVGan replied:
Added

JAVGan commented:
@johnbieren Any reason the task  test failed?

I'll send the gitlint fix in a new commit

JAVGan commented:
I've fixed the gitlint but now I'm not sure why the `test-marketplacesvm-push-disk-images` fails.. @johnbieren thoughts?

JAVGan commented:
> I've fixed the gitlint but now I'm not sure why the `test-marketplacesvm-push-disk-images` fails.. @johnbieren thoughts?

I think I've got a clue... In the `schema` the `starmap` and `productInfo` should be within the `.components` section and in the task/tests it was not...

I've fixed it and tried to simplify things as well

johnbieren commented on tasks/marketplacesvm-push-disk-images/marketplacesvm-push-disk-images.yaml:
```diff
@@ -0,0 +1,143 @@
+
+        # Setup required variables
+        SNAPSHOT_JSON=$(jq -c '.' "$(workspaces.data.path)/$(params.snapshotPath)")
+        STARMAP_MAPPING=$(jq -c '.components[0].starmap' <<< "$SNAPSHOT_JSON")
```
Why component 0? What if there is more than one component?

JAVGan replied:
Ideally there should be only "one" mapping with a list of everything...

So... Should I move it outside the components? Just in `mappings` like the secret?

johnbieren replied:
If it isn't going to change per component, I wouldn't have it nested inside component, yeah. It can be inside mapping like the secret or a new starMap top level key if you'd prefer

JAVGan replied:
Nevermind... I've updated it so it will append all the component's mappings into a single array. Then we can be more flexible 😄

johnbieren replied:
If you say so :man_shrugging: 
`Ideally there should be only "one" mapping with a list of everything` makes it sound to me like it should not be inside the component object, but you know this stuff better than me

johnbieren commented on tasks/marketplacesvm-push-disk-images/marketplacesvm-push-disk-images.yaml:
```diff
@@ -0,0 +1,143 @@
+        # Setup required variables
+        SNAPSHOT_JSON=$(jq -c '.' "$(workspaces.data.path)/$(params.snapshotPath)")
+        STARMAP_MAPPING=$(jq -c '.components[0].starmap' <<< "$SNAPSHOT_JSON")
+        STARMAP_MAPPING_FILE=$(workspaces.data.path)/starmap.yaml
```
This isn't a good idea. It will cause issues with concurrent runs if the same workspace is used. I'd recommend something like this instead https://github.com/konflux-ci/release-service-catalog/blob/development/tasks/create-pyxis-image/create-pyxis-image.yaml#L94

JAVGan replied:
Ok!

johnbieren commented on tasks/marketplacesvm-push-disk-images/tests/test-marketplacesvm-push-disk-images.yaml:
```diff
@@ -0,0 +1,190 @@
+              }
+              EOF
+
+              cat > "$(workspaces.data.path)/starmap.json" << EOF
```
This file isn't passed to the task. Why is it here?

JAVGan replied:
Just to compare the expected results in the tests

JAVGan commented:
```
task run-task has failed: Failed to create pod due to config error
```

it seems to fail right after the 2 `cat` commands on setup...

johnbieren commented on tasks/marketplacesvm-push-disk-images/tests/test-marketplacesvm-push-disk-images.yaml:
```diff
@@ -0,0 +1,190 @@
+        - name: snapshotPath
+          value: "snapshot_spec.json"
+        - name: cloudMarketplacesSecret
+          value: test-secret
```
This is the wrong value

JAVGan replied:
yes, caught that, thanks!

johnbieren commented on tasks/marketplacesvm-push-disk-images/tests/test-marketplacesvm-push-disk-images.yaml:
```diff
@@ -0,0 +1,190 @@
+                        "revision": "1abbfcdbc1c5e8b7aba07673297237ed192b50e2",
+                        "url": "https://gitlab.com/konflux/test-product/disk-images/amd-bootc"
+                      }
+                    }
```
missing comma here I think? A comma is missing somewhere... that is why it is saying ` parse error: Expected separator between values at line 14, column 14`

JAVGan replied:
Thank! Fixed

JAVGan commented:
Hmmmmm....

It seems my regex from https://github.com/konflux-ci/release-service-utils/pull/326/files is not happy with a dot in the dirname

```
[run-task : pull-and-push-images-to-marketplaces]   File "/home/pubtools-marketplacesvm-wrapper/marketplacesvm_push_wrapper", line 149, in <module>
[run-task : pull-and-push-images-to-marketplaces]     main()
[run-task : pull-and-push-images-to-marketplaces]   File "/home/pubtools-marketplacesvm-wrapper/marketplacesvm_push_wrapper", line 113, in main
[run-task : pull-and-push-images-to-marketplaces]     args = validate_env_vars(parse_args())
[run-task : pull-and-push-images-to-marketplaces]   File "/home/pubtools-marketplacesvm-wrapper/marketplacesvm_push_wrapper", line 108, in validate_env_vars
[run-task : pull-and-push-images-to-marketplaces]     args.source = get_source_url(args.source)
[run-task : pull-and-push-images-to-marketplaces]   File "/home/pubtools-marketplacesvm-wrapper/marketplacesvm_push_wrapper", line 76, in get_source_url
[run-task : pull-and-push-images-to-marketplaces]     raise ValueError("Not a valid staging directory: %s" % item)
[run-task : pull-and-push-images-to-marketplaces] ValueError: Not a valid staging directory: /tmp/tmp.Vde3yHJTof
```

JAVGan commented:
@johnbieren mind if I open a fix PR on https://github.com/konflux-ci/release-service-utils ?

johnbieren commented:
> @johnbieren mind if I open a fix PR on https://github.com/konflux-ci/release-service-utils ?

I think that is the best solution. Then you can bump the utils image here once it is merged and available

JAVGan commented:
> > @johnbieren mind if I open a fix PR on https://github.com/konflux-ci/release-service-utils ?
> 
> I think that is the best solution. Then you can bump the utils image here once it is merged and available

Thanks!

There it goes: https://github.com/konflux-ci/release-service-utils/pull/338

johnbieren commented:
The on push pipeline for utils failed for some flake (or outage). I will let you know when I see the new image available

johnbieren commented:
New image available `quay.io/konflux-ci/release-service-utils:6556e8a6b031c1aad4f0472703fd121a6e1cd45d`
(and this needs rebase)

JAVGan commented:
> New image available `quay.io/konflux-ci/release-service-utils:6556e8a6b031c1aad4f0472703fd121a6e1cd45d` (and this needs rebase)

Thanks! I'll update it

JAVGan commented:
@johnbieren rebased and updated

JAVGan commented:
The `starmap.yaml` was no longer generated into `$(workspaces.data.path)` so it couldn't be found. I modified the `mock.sh` to pick it and generate the file where the test expect

JAVGan commented:
Hmm it seems it get deleted due to the pod shutdown.

@johnbieren which should be better:

1. Remove the test to ensure starmap mapping is the same as input
2. Bring back `starmap.yaml` to workspace instead of using ephemeral storage

johnbieren commented:
Still `starmap.yaml` errors

JAVGan commented on tasks/marketplacesvm-push-disk-images/marketplacesvm-push-disk-images.yaml:
```diff
@@ -0,0 +1,143 @@
+        # Setup required variables
+        SNAPSHOT_JSON=$(jq -c '.' "$(workspaces.data.path)/$(params.snapshotPath)")
+        STARMAP_MAPPING=$(jq -c '[.components[].starmap[]]' <<< "$SNAPSHOT_JSON")
+        STARMAP_MAPPING_FILE="$(dirname "$(params.snapshotPath)")/starmap.yaml"
+        yq -p json -o yaml <<< "$STARMAP_MAPPING" > "$STARMAP_MAPPING_FILE"
+
```
@johnbieren it seems `$(dirname "$(params.snapshotPath)")` results in an ephemerous storage... Is it ok to bring it back to `$(workspaces.data.path)`?

johnbieren replied:
I don't really understand why
```
  starmap_file="${@: -1}"
  cat "$starmap_file" > "$(workspaces.data.path)/starmap.yaml"
```
is in your mocks.sh. Isn't the yaml generated by this code right here on line 47?

JAVGan replied:
I was trying to pick the file after the task ended. But it seems it deletes the file when the task ends, as it's stored in `$(dirname "$(params.snapshotPath)")/starmap.yaml`.

My goal on the `mocks` was to write in the `workspaces.data` so it would be preserved but it seems it's gone before

JAVGan replied:
basically:
-  `.starmap.yaml` missing after task finishes
- no data stored in worspaces.data

JAVGan replied:
I've pushed again using the `workspace.data` within the task, let's see if it works now

johnbieren replied:
I mean it doesn't really matter if it works with the path you changed it back to... I can't/won't approve it with that path. It breaks parallelism

johnbieren replied:
> My goal on the mocks was to write in the workspaces.data so it would be preserved but it seems it's gone before

The task is writing it. Why are you also writing it in mocks.sh?

johnbieren replied:
Use `$(workspaces.data.path)/$(dirname "$(params.snapshotPath)")/starmap.yaml`

JAVGan replied:
Ok, updating

JAVGan replied:
Updated🤞🏻

johnbieren reviewed (dismissed):
/ok-to-test

johnbieren commented on tasks/marketplacesvm-push-disk-images/tests/test-marketplacesvm-push-disk-images.yaml:
```diff
@@ -0,0 +1,190 @@
+                exit 1
+              fi
+
+              # Single component fixture: Expected "prepare_componentonent" to be called once
```
typo componentonent

JAVGan replied:
Fixed, thanks!

johnbieren reviewed (dismissed):
/ok-to-test

JAVGan commented:
@johnbieren I'll start working in the pipeline... Feel free to merge this whenever it completes all steps

johnbieren commented:
> @johnbieren I'll start working in the pipeline... Feel free to merge this whenever it completes all steps

Sounds good. I plan to wait for Martin to get a chance to take a look tomorrow in case he has any other comments, then we will merge

johnbieren commented:
/ok-to-test

johnbieren commented:
E2E irrelevant here. Will merge tomorrow unless @mmalina comments something to change

mmalina commented on tasks/marketplacesvm-push-disk-images/marketplacesvm-push-disk-images.yaml:
```diff
@@ -0,0 +1,143 @@
+              key: key
+      script: |
+        #!/usr/bin/env bash
+        set -ex
```
Is there a reason this one (and the mock script) doesn't use `set -eux`? I think it's nice to have `-u`, because ideally you shouldn't access variables that are undefined. But I'm not saying it's strictly required, so it's up to you. Also, if you've already tested this and changing this would require more testing, I understand you'd probably want to keep it as is.

JAVGan replied:
I'll patch it up to `eux`, just a sec

JAVGan replied:
@mmalina I've updated it to `eux`

JAVGan replied:
Thanks for this suggestion @mmalina !!

In fact it detected an unbound variable!

lslebodn replied:
@JAVGan I would even recommend to enable `-o pipefail` (there is not a short version)

man bash says
```
                      pipefail
                              If set, the return value of a pipeline is the value of the last (right‐
                              most) command to exit with a non-zero status, or zero if  all  commands
                              in the pipeline exit successfully.  This option is disabled by default.
```
https://gist.github.com/mohanpedala/1e2ff5661761d3abd0385e8223e16425

JAVGan replied:
@lslebodn I would prefer to keep it simpler and streamlined with other tasks...

mmalina replied:
Yeah, I agree `-o pipefail` is nice in most cases as well. But probably because it doesn't have a shorthand version, it's not used as much. I'm fine with the current version now.

lslebodn replied:
@mmalina
it is poor excuse to not use it in Konflux  due to missing short version. :-)
one can easily combine sort and long variants without any problem
`set -euo pipefail`

mmalina replied:
It's not really an excuse, I was just trying to find explanation for why it's not used as much (not only in our repo, but also the default tekton uses for scripts is `set -eux`) ;)

I agree it's nice and I will definitely consider using it next time I create a script.

johnbieren replied:
We have it in 9 of our tasks. It isn't not used in Konflux. Really we just ask the task script and tests/mocks file to have the same setting for testing purposes. Being consistent across tasks would be nice yeah but we have many people contributing tasks and it is kind of up to them what options they want and think make sense

johnbieren reviewed (dismissed):
/ok-to-test

JAVGan commented:
@johnbieren @mmalina Fixed an unbound variable

johnbieren commented:
/ok-to-test

mmalina commented on tasks/marketplacesvm-push-disk-images/tests/mocks.sh:
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/env bash
+set -ex
```
Now that the main script was updated to `set -eux`, this one should probably use it as well.

JAVGan replied:
Patched 😄

mmalina replied:
Thanks :)

mmalina reviewed (commented):
One last small comment and then it should be ready to merge.
